### DISCUSSION TITLE
How to create instance of, set properties, and attach Na-channel mechanism?

### DISCUSSION THREAD

**Discussion by lungsi** [arbor-sim/arbor#1150] [General] [answered]
Based on the [documentation](https://arbor.readthedocs.io/en/pydoc/mechanisms.html?#mechanism.values) I can create an instance of the leaky-channel with
```
leaky_chnl = arbor.mechanism('passive/el=-45', {'g': 0.1})
```
and create instance for [Hodgkin-Huxley based mechanism](https://arbor.readthedocs.io/en/pydoc/mechanisms.html?#density-mechanisms) with
```
arbor.mechanism('hh')
```

Then, how does one do it for Hodgkin-Huxley based Na-channel? [Documentation for NMODL mechanisms](https://arbor.readthedocs.io/en/pydoc/nmodl.html#ions) says
> Arbor recognizes na, ca and k ions by default.

Does that mean I can achieve it with
```
na_chnl = arbor.mechanism('hh/na/el=-45', {'g': 0.1})
```
or is it?
```
na_chnl = arbor.mechanism('na/el=-45', {'g': 0.1})
```

PS: If one of the above (two) is correct, I can't tell which is the correct way because I can run both lines without any error or warning messages.

**Comment by halfflat** [comment, marked as answer]
We probably should be giving an error on those calls to `arbor.mechanism` — they won't be doing anything useful!

When adding new mechanisms defined in NMODL — which is still an awkward process in Arbor, unfortunately — any ion name can be used. But these ions have to be declared, along with their charge, in the cable cell global information. For convenience, `na`, `ca`, and `k` are already set up.

The HH mechanism that is available is modelling three current sources in the one mechanism: a passive current; a sodium-channel current; and a potassium-channel current. While there is a 'pas' mechanism that just models the passive current, we do not have mechanisms that model just the HH sodium channel or the HH potassium channel. But we do see that parameterized mechanisms for individual ion channels is simply a useful thing to have. I will make an issue regarding this.

The way the HH mechanism NMODL has been written prevents setting the steady-state potential for the leak current with the syntax `hh/el=-45`; rather, it has to be set as one of the parameters in the dictionary, e.g.
```
arbor.mechanism('hh', {'el': -45})
```

But this really shouldn't be so. We (and NMODL) make a distinction between 'global parameters' and 'range parameters'. Global parameters have one value for the whole mechanism, while range parameters can change with location. When we add a mechanism `passive/el=-45`, a _new version_ of the mechanism `passive` is created with the name `passive/el=-45` that has the global parameter `el` set to -45. The conductance `g` though is a 'range parameter', and so the same mechanism can have different `g` values at different parts of the morphology. Range parameters, for mathematical reasons, should really be limited to conductances and channel densitities, but the mechanisms that Arbor ships with do not consistently follow this convention, and the HH mechanism is one of these. Again, I'll raise this as an issue.

This is a very long-winded explanation for why `hh/el=-45` doesn't work, but `passive/el=-45` does. There is a more thorough explanation at https://arbor.readthedocs.io/en/pydoc/mechanisms.html#parameters .

If though, at the end of the day, the goal is to use the whole HH model (leak, sodium and potassium currents) but change the reversal potential used for the sodium channel, then this is achieved in a different manner again: ionic reversal potentials, be they constant or computed via the Nernst equation, are associated with descriptions of ion state rather than in mechanisms.
There is some discussion of this in https://arbor.readthedocs.io/en/pydoc/cable_cell.html .

**Reply by lungsi** [reply]
Thank you @halfflat.

Its good to know that although `"passive"` and `"hh"` are names of `arbor.mechanism`, to change the reverse potential `el` of the mechanism the syntax is different
```
arbor.mechanism("passive/el=-45")
arbor.mechanism("hh", {"el": -45"})
```
[PS: I think there is a typo in your expression `arbor.mechanism('hh', {'el': =45})`. I think the correct expression should be `arbor.mechanism('hh', {'el': 45})`]

As you have clearly explained that the above two expressions results in creating new mechanisms with respective global parameter (`"el"`),
and in terms of Arbor syntax since arguments for `arbor.mechanism` is  configured as `(String, Dictionary)` such that to simultaneously
set the range/`"g"` parameter for `passive` mechanism one does it with `arbor.mechanism("passive/el=-45", {"g": 0.01})` then,

"How does one set the range parameter for `hh` mechanism? Is it `arbor.mechanism("hh", {"el": -45, "g": 0.01})`? (but then, if `"hh"` is a generic HH mechanism I guess we can't tell whether the mechanism being set is for Na-channel or some other ion channel)"

Also, when you say
> The HH mechanism that is available is modelling three current sources in the one mechanism: a passive current; a sodium-channel current; and a potassium-channel current.

but at the same time
> ... we do not have mechanisms that model just the HH sodium channel or the HH potassium channel.

presumably because as you say
> For convenience, na, ca, and k are already set up.

then, what is an example of using the `"hh"` mechanism?

Finally since,
> For convenience, na, ca, and k are already set up.

and based on the [documentation for setting up ion species](https://arbor.readthedocs.io/en/pydoc/cable_cell.html#ion-species) one can either
* `cell.set_ion(ion = "na", method = arbor.mechanism("nernst/x=na"))`
* `cell.set_ion(ion = "na", method = "nernst/x=na")`
* `cell.paint("soma", arbor.ion("na", rev_pot=-45)`
* `cell.paint("soma", arbor.ion("na", int_con=2e-4, ext_con=2.5, rev_pot=-45))`

where documentation for the `arbor.ion` says
>> For setting ion properties (internal and external concentration [mM] and reversal potential [mV]) on cells and regions.

**Is it not possible to set the range/`"g"` parameter using the in-built `arbor.ion`?** [setting the range/`"g"` parameter either for the whole cell or for specific cell regions]

**Reply by halfflat** [reply]
Thanks for pointing out the typo! I've gone and corrected it.

Regarding mechanism parameter setting, for any mechanism that has (for example) `foo` declared as a range parameter in the NMODL source, that value can be set locally with an entry like `{'foo': 123}` in the dictionary parameter to `arbor.mechanism`. To be honest, the name `arbor.mechanism` is a bit misleading: it's really a combination of a mechanism (a biophysical process typically described in NMODL with a fixed set of global paramters) and a set of values for local parameters to be used when it is painted on a cell.

When we write `arbor.mechanism("hh", {"el": -45})`, we are saying we want the mechanism with the name "hh", with one of its local parameters "el" set to -45. Ultimately what "el" means is up to the mechanism "hh", and depends upon the underlying NMODL description.

For the "hh" mechanism used in Arbor (and also, in NEURON), the "el" parameter strictly means the reversal potential for the passive leak current.

If one were to look at the NMODL definition of the "hh" mechanism ( https://github.com/arbor-sim/arbor/blob/master/mechanisms/default/hh.mod ), one would see that it has the statement `USEION na READ ena WRITE ina` near the top. This means that the mechanism is getting the reversal potential `ena` for the sodium ion current from the data associated with the ion `na`. Generally speaking, if more than one mechanism wants to use the sodium ion reversal potential, they should all use the same value, and a statement like this `USEION na READ ena` in a mechanism definition means that the mechanism will use this common value of `ena`.

As you've seen in the docs, ion data can be set for the whole cell, and then overridden on parts of the cell as required. Reversal potentials are a special case: you can use both `set_ion` and `paint` to specify an ion's reversal potential on all or part of a cell, but if you want a potential that is dynamic (i.e. depends upon time-varying internal and external ion concentrations), then you can specify a method for computing that only over the whole cell, with `cell.set_ion(..., method="...")`. The 'method' is the name of a mechanism that does the work of computing the reversal potential every time step.

On the other hand, the `g` values correspond to total conductance coefficients in mechanisms like `pas` and `hh`. The `hh` mechanism has three conductance parameters that can be set: `gl` for the leak current (this corresponds to the parameter `g` in `pas`); `gna` for the sodium ion channel current coefficient; and `gk` for the potassium ion channel current coefficient. The ion channel conductance parameters are essentially the product of the per-ion channel conductance, and the density of that ion channel on the membrane. The actual specific conductance at any point in time will be the product of (e.g.) `gna` and the proportion of open sodium ion channels as governed by the HH-style dynamics.

**Comment by lungsi** [comment]
> The `hh` mechanism has three conductance parameters that can be set

**How do I set these parameters?**

Also, with regards to the conductance parameters so that we mean the same, given the expression

<img src="http://latex.codecogs.com/svg.latex?C\frac{dV_m}{dt}&space;=&space;-G_{Na}\left&space;(&space;V_m&space;-&space;E_Na&space;\right&space;)&space;-G_{K}\left&space;(&space;V_m&space;-&space;E_K&space;\right&space;)&space;-G_{lk}\left&space;(&space;V_m&space;-&space;E_{lk}&space;\right&space;)" title="http://latex.codecogs.com/svg.latex?C\frac{dV_m}{dt} = -G_{Na}\left ( V_m - E_Na \right ) -G_{K}\left ( V_m - E_K \right ) -G_{lk}\left ( V_m - E_{lk} \right )" />

where,

<img src="http://latex.codecogs.com/svg.latex?G_x&space;=&space;g_x&space;m^p&space;h^q" title="http://latex.codecogs.com/svg.latex?G_x = g_x m^p h^q" />

such that,
| _x_  | _p_   | _q_ |
| :---: | :----: | :--: |
|  Na  | 3      | 1   |
|  K    | 4      | 0   |
|  lk   | 0      | 0   |

[NB: <i>m</i><sup>4</sup> = <i>n</i><sup>4</sup>]

Looking at [`hh.mod`](https://github.com/arbor-sim/arbor/blob/master/mechanisms/default/hh.mod) I am guessing `gl`, `gna`, and `gk` correspond to the above equations as
*  `gna` &rarr; <i>G<sub>Na</sub></i>
*  `gk` &rarr; <i>G<sub>K</sub></i>
*  `gl` &rarr; <i>G<sub>lk</sub></i>

and
*  `gnabar` &rarr; <i>g<sub>Na</sub></i>
*  `gkbar` &rarr; <i>g<sub>K</sub></i>

Since `gl` in `passive` mechanism is set using `g` as
```
arbor.mechanism('passive/el=-45', {'g': 0.1})
```
For `hh` mechanism (based on [`hh.mod`](https://github.com/arbor-sim/arbor/blob/master/mechanisms/default/hh.mod)) I guess to set the conductance parameters the keys will be: `gl`, `gnabar` and `gkbar`.

**How do I set them?**

Since,  documentation for `arbor.ion` says

> For setting ion properties (internal and external concentration [mM] and reversal potential [mV]) on cells and regions.

Do I set the parameters `gl`, `gnabar` and `gkbar` with
```
arbor.mechanism('hh', {'el': 45, 'gl': 0.0003, 'gkbar': 0.036, 'gnabar': 0.120})
```
Also, if one created two `arbor.mechanism`; `passive` and `hh` as below
```
leaky_chnl = arbor.mechanism('passive/el=-54.3', {'g': 0.0003})
hh_na_k_chnl = arbor.mechanism('hh', {'ek': -77., 'ena': 50., 'gkbar': 0.036, 'gnabar': 0.120})
```
then to attach `arbor.mechanism` to particular regions of the cell one must _paint_, i.e
```
cell.paint('soma', hh_na_k_chnl)
```
my question then is **How do I attach the above `leaky_chnl` mechanism to not just a region of the cell but to the whole cell?**

Do I have to paint the `leaky_chnl` mechanim by looping over all the regions of the cell?

**Reply by halfflat** [reply]
You have it exactly right regarding the `hh` parameters!

Just as `'soma'` is a label describing part of the morphology, we can make another label that describes all of it. This is done in the label dictionary used when making the cell, e.g.
```
cell = arbor.cable_cell(morph, arbor.label_dict({'soma': '(tag 1)', 'everywhere': '(all)'}))
```

Then the dynamics would be applied with
```
cell.paint('soma', hh_na_k_chnl)
cell.paint('everywhere', leaky_chnl)
```

Note though that this is essentially putting two different leaky channels on the soma, because `hh` includes one too. Our little language for describing parts of the morphology will improve in the next release of Arbor, but for now the easiest way to work around this is probably to set the leak conductance to zero in the `hh` mechanism, i.e.
```
hh_na_k_chnl = arbor.mechanism('hh', {'ek': -77., 'ena': 50., 'gkbar': 0.036, 'gnabar': 0.120, 'gl': 0.0})
```

Regarding the new release (0.4): we're aiming to have that published within the next two weeks, and some of the changes to the API will hit our master branch before then, which will probably mean that some changes to your python code will be required too I'm afraid.

**Reply by lungsi** [reply]
Thank you.

> Regarding the new release (0.4): we're aiming to have that published within the next two weeks, and some of the changes to the API will hit our master branch before then, which will probably mean that some changes to your python code will be required too I'm afraid.

I will have to pin this 🥇

**Comment by lungsi** [comment]
I have another curious (sort-off unrelated) question.

In the expression below for painting the ionic properties
```
cell.paint("soma", arbor.ion("na", int_con=2e-4, ext_con=2.5, rev_pot=-45))
```
and since

<img src="http://latex.codecogs.com/svg.latex?V_{Na}=\frac{kT}{e}ln\left(\frac{[Na^&plus;]_o}{[Na^&plus;]_i}\right)&space;&plus;&space;\frac{kT}{e}ln\left(\frac{1&space;&plus;[\Delta&space;Na^&plus;]/[Na^&plus;]_o}{1-[\Delta&space;Na^&plus;]/[Na^&plus;]_i}\right)" title="http://latex.codecogs.com/svg.latex?V_{Na}=\frac{kT}{e}ln\left(\frac{[Na^+]_o}{[Na^+]_i}\right) + \frac{kT}{e}ln\left(\frac{1 +[\Delta Na^+]/[Na^+]_o}{1-[\Delta Na^+]/[Na^+]_i}\right)" />

where there Nernst (or reverse) potential is

<img src="http://latex.codecogs.com/svg.latex?E_{Na}=\frac{kT}{e}ln\left(\frac{[Na^&plus;]_o}{[Na^&plus;]_i}\right)" title="http://latex.codecogs.com/svg.latex?E_{Na}=\frac{kT}{e}ln\left(\frac{[Na^+]_o}{[Na^+]_i}\right)" />

then, because the value for the `rev_pot` is already supplied what does `int_con` and `ext_con` correspond to?

For instance, is `int_con` equal to?
* <img src="http://latex.codecogs.com/svg.latex?[Na^&plus;]_i" title="http://latex.codecogs.com/svg.latex?[Na^+]_i" />
* or is it equal to <img src="http://latex.codecogs.com/svg.latex?\frac{[\Delta&space;Na^&plus;]}{[Na^&plus;]_i}" title="http://latex.codecogs.com/svg.latex?\frac{[\Delta Na^+]}{[Na^+]_i}" />

If it is the first why pass the value of <i>E<sub>Na</sub></i> (`rev_pot`)?
